tests-e2e/LocalOcw.ts

```diff
@@ -0,0 +1,211 @@
+import * as path from "node:path"
+import * as http from "node:http"
+import { SpawnOptions } from "node:child_process"
+import execShCb from "exec-sh"
+import handler from "serve-handler"
+import Table from "cli-table3"
+import * as color from "ansi-colors"
+
+import {
+  TEST_SITES,
+  LOCAL_OCW_PORT,
+  siteUrl,
+  TestSiteAlias
+} from "./util/test_sites"
+import SimpleServer, { RedirectionRule } from "./util/SimpleServer"
+import { env, cleanEnv, stringifyEnv } from "../env"
+import { hugo } from "../package_scripts/util"
+
+const execSh = execShCb.promise
+
+const fromRoot = (...pathFromRoot: string[]) => {
+  return path.join(__dirname, "../", ...pathFromRoot)
+}
+
+/**
+ * Redirect all requests /path/to/thing to /ocw-ci-test-www/path/to/thing,
+ * except requests to /courses/.
+ */
+const OCW_WWW_REDIRECT: RedirectionRule = {
+  type:      "rewrite",
+  match:     /^\/(?!(courses\/))/,
+  transform: url => `/ocw-ci-test-www${url}`
+}
+
+interface LocalOCWOptions {
+  /**
+   * All sites will be built to subdirectories of this directory.
+   */
+  rootDestinationDir: string
+  staticApiPort: number
+}
+
+/**
+ * Helps build and serve OCW sites locally.
+ */
+class LocalOCW {
+  private rootDestinationDir: string
+
+  private staticApiPort: number
+
+  /**
+   * A server that responds to STATIC_API_BASE_URL requests with JSON from
+   * test-sites/__fixtures__.
+   */
+  staticApiServer: {
+    /**
+     * Set a handler to be called before the static fixtures handler.
+     */
+    patchHandler: (handler: http.RequestListener) => void
+    /**
+     * Reset the handler; i.e., clears `patchHandler`
+     */
+    resetHandler: () => void
+    /**
+     * Close the server
+     */
+    close: () => void
+    /**
+     * Listen to requests
+     */
+    listen: () => void
+  }
+
+  constructor({ rootDestinationDir, staticApiPort }: LocalOCWOptions) {
+    this.rootDestinationDir = rootDestinationDir
+    this.staticApiPort = staticApiPort
+    this.staticApiServer = this.makeStaticApiServer()
+  }
+
+  private makeStaticApiServer = () => {
+    let patchedHandler: http.RequestListener | undefined
+    const server = new SimpleServer(
+      (request, response) => {
+        if (patchedHandler) {
+          patchedHandler(request, response)
+          if (response.headersSent) return
+        }
```

Comment:
This is the new code that was not in `global-setup.ts` before, our playwright test setup file.

One of the primary things to test with build failures is how our hugo theme reacts to static api request failures. `patchedHandler` let's us override a static api request for a single test.